README.md
# torotator

SOCKS5 proxy that balances across several Tor sessions.

This project may be used to create one or more Tor sessions, providing access
to the Tor network as a sort of load balancer using HAProxy. Each Tor instance
is rotated after a certain amount of time, and each Tor circuit is re-routed
periodically as well.


CHANGELOG
# Version 0.2.0 - 2016.12.27

This version removes the HTTP proxy piece about torotator and simply load
balances and rotates SOCKS5 proxies directly. Privoxy has been removed.

If an HTTP proxy is required, it's far easier to create one instance of privoxy
or whatever proxy you prefer in front of the load balanced, rotating SOCKS5
proxies provided by torotator.


cmd/cmd.go
package main

import (
	"bufio"
	"context"
	"errors"
	"io"
	"os/exec"
	"time"

	"github.com/uber-go/zap"
)

// Cmd is a wrapper around exec.Cmd. It allows for stdout and stderr to automatically be logged along with everything
// else in the application. It also provides helpers to check if the process has finished and also to clean up the
// process.
type Cmd struct {
	log    zap.Logger
	cmd    *exec.Cmd
	stdout io.ReadCloser
	stderr io.ReadCloser
	done   chan struct{}

	transformLog func(string) (string, string, []zap.Field)
}

// NewCommand creates a new Cmd that is setup for common logging and state tracking.
func NewCommand(ctx context.Context, log zap.Logger, name string, args ...string) (c *Cmd, err error) {
	c = &Cmd{
		log:  log,
		cmd:  exec.CommandContext(ctx, name, args...),
		done: make(chan struct{}),
	}

	if c.stdout, err = c.cmd.StdoutPipe(); err != nil {
		c.log.Error("failed to setup stdout pipe", zap.Error(err))
	}

	if c.stderr, err = c.cmd.StderrPipe(); err != nil {
		c.log.Error("failed to setup stderr pipe", zap.Error(err))
	}

	if err = c.cmd.Start(); err != nil {
		c.log.Error("failed to start", zap.Error(err))
		return nil, err
	}

	c.log = c.log.With(zap.Int("pid", c.cmd.Process.Pid))

	// give the process a bit of time to settle
	time.Sleep(250 * time.Millisecond)

	// only ended processes have a non-nil ProcessState
	if c.cmd.ProcessState != nil {
		return nil, errors.New(c.cmd.ProcessState.String())
	}

	c.log.Info("running")

	return c, nil
}

// Pid returns the PID of the underlying command.
func (c *Cmd) Pid() int {
	if c.cmd == nil {
		return -1
	}

	return c.cmd.Process.Pid
}

// Done returns a channel that signals when the process has ended.
func (c *Cmd) Done() <-chan struct{} {
	return c.done
}

// Wait processes output from the process and signals when the process has neded.
func (c *Cmd) Wait() {
	var (
		line   string
		fields []zap.Field
		level  string
		lf     func(string, ...zap.Field)
	)

	// receive data from both stdout and stderr
	r := io.MultiReader(c.stdout, c.stderr)

	// wait for output
	scanner := bufio.NewScanner(r)
	for scanner.Scan() {
		// extract log level information from Tor messages
		line = scanner.Text()
		fields = fields[:]

		// optionally process output from the command to make common logging more useful
		if c.transformLog != nil {
			level, line, fields = c.transformLog(line)
		}

		switch level {
		case "debug":
			lf = c.log.Debug
		case "warn":
			lf = c.log.Warn
		case "err", "fatal":
			lf = c.log.Error
		default:
			lf = c.log.Info
		}

		lf(line, fields...)
	}

	if err := scanner.Err(); err != nil {
		c.log.Error("output error", zap.Error(err))
	}

	// wait for the underlying process to finish
	c.cmd.Wait()

	// signal that the command has ended
	close(c.done)
}

// Close does its best to clean up the process.
func (c *Cmd) Close() (err error) {
	// presence of ProcessState means the process has already exited
	if c.cmd.ProcessState != nil {
		return nil
	}

	c.log.Debug("killing process")
	if err = c.cmd.Process.Kill(); err != nil {
		return
	}

	if c.cmd.ProcessState == nil {
		c.log.Debug("waiting for process to exit")
		if err = c.cmd.Wait(); err != nil {
			return
		}
	}

	return nil
}


cmd/tor.go
package main

import (
	"context"
	"fmt"
	"os"
	"path"
	"strings"
	"time"

	"github.com/uber-go/zap"
)

type Tor struct {
	log  zap.Logger
	cmd  *Cmd
	port int
	dir  string
	pid  string
}

func NewTor(ctx context.Context) (t *Tor, err error) {
	t = &Tor{}

	// loop until we find a port we like
	for {
		select {
		case <-ctx.Done():
			return nil, fmt.Errorf("application terminating")
		default:
		}

		t.port = portPlz()
		t.log = log.With(zap.String("service", "tor"), zap.Int("port", t.port))
		t.dir = fmt.Sprintf("/tmp/torotator/tor-%d", t.port)
		t.pid = path.Join(t.dir, "tor.pid")

		t.MakeDirs()

		t.cmd, err = NewCommand(ctx, t.log, "tor",
			"--allow-missing-torrc",
			"--SocksPort", fmt.Sprintf("%d", t.port),
			"--NewCircuitPeriod", fmt.Sprintf("%d", *circuitTime),
			"--DataDirectory", t.dir,
			"--PidFile", t.pid,
			"--Log", "warn stdout")
		if err != nil {
			t.log.Error("failed to setup command", zap.Error(err))
			time.Sleep(500 * time.Millisecond)
			continue
		}

		t.cmd.transformLog = t.TorLogger

		break
	}

	return t, nil
}

func (t *Tor) MakeDirs() (err error) {
	if err = os.MkdirAll(t.dir, 0700); err != nil {
		return
	}

	return nil
}

func (t *Tor) TorLogger(line string) (level, msg string, fields []zap.Field) {
	line = line[21:]
	lvlPos := strings.Index(line, "]")
	level = line[:lvlPos]
	msg = line[lvlPos+2:]

	return
}

func (t *Tor) Done() <-chan struct{} {
	return t.cmd.Done()
}

func (t *Tor) Wait() {
	t.cmd.Wait()
}

func (t *Tor) Close() (err error) {
	if t == nil {
		return nil
	}

	defer func() {
		if err = os.RemoveAll(t.dir); err != nil {
			t.log.Error("failed to remove data directory", zap.String("path", t.dir), zap.Error(err))
		}
	}()

	t.cmd.log.Info("cleaning up")
	if err = t.cmd.Close(); err != nil {
		if err.Error() != "signal: killed" {
			t.cmd.log.Error("failed to kill server", zap.Error(err))
		}
		return
	}

	return nil
}


cmd/rotate.go
package main

import (
	"context"
	"os"
	"os/signal"
	"sync"
	"time"

	"github.com/uber-go/zap"
)

const (
	VERSION          = "0.1.0"
	TOR_COUNT        = 3
	PORT_RANGE_START = 30000
	HAPROXY_CFG      = "/etc/haproxy.cfg"
	MAX_PROXY_TIME   = 900
	TEST_URL         = "http://echoip.com"
)

var (
	log zap.Logger
	//haproxy Service
)

func main() {
	ports = make(map[uint]uint)

	log = zap.New(zap.NewJSONEncoder(zap.RFC3339Formatter("time")))
	log.Info("rotating tor proxy", zap.String("version", VERSION))

	ctx := SignalContext()
	wg := new(sync.WaitGroup)

	ha, err := NewHAProxy(ctx, 8080)
	if err != nil {
		log.Fatal("failed to start HAproxy", zap.Error(err))
	}
	go ha.Wait()

	Rotate(ctx, wg, ha)

	// clean up
	wg.Wait()
	log.Info("done")
}

func Rotate(ctx context.Context, wg *sync.WaitGroup, ha *HAProxy) {
	// Used to limit the number of running proxies. This is separate from wg because wg is unbounded.
	c := make(chan bool, TOR_COUNT)

	for {
		select {
		case <-ctx.Done():
			// application terminating
			close(c)
			return

		default:
			c <- true
			wg.Add(1)
			go func() {
				RunProxy(ctx, ha)

				wg.Done()
				<-c
			}()
		}
	}
}

func RunProxy(ctx context.Context, ha *HAProxy) {
	// create a new tor/privoxy pair
	tor, err := NewTor(ctx)
	if err != nil {
		tor.Close()
		return
	}

	privoxy, err := NewPrivoxy(ctx, tor)
	if err != nil {
		tor.Close()
		privoxy.Close()
		return
	}

	// mark the ports as used
	mapPorts(tor.port, privoxy.port)

	_log := log.With(zap.Uint("tor", tor.port), zap.Uint("privoxy", privoxy.port))
	_log.Info("proxy started")

	ha.AddBackend(privoxy.port)

	// let the processes run until they terminate
	go tor.Wait()
	go privoxy.Wait()

	// TODO periodically check that this proxy is still functional
	// wait for any of the following events to occur
	select {
	case <-ctx.Done():
		// application terminating
	case <-tor.Done():
		// tor ended
	case <-privoxy.Done():
		// privoxy ended
	case <-time.After(time.Duration(MAX_PROXY_TIME) * time.Second):
		// proxy lifetime expired
	}

	// clean up after ourselves
	_log.Info("stopping proxy")
	privoxy.Close()
	tor.Close()

	unmapPorts(tor.port, privoxy.port)
	_log.Info("proxy terminated")
}

func SignalContext() context.Context {
	ctx, cancel := context.WithCancel(context.Background())

	// handle termination signals
	terminate := make(chan os.Signal, 1)
	signal.Notify(terminate, os.Kill, os.Interrupt)

	go func() {
		<-terminate
		cancel()
	}()

	return ctx
}


cmd/haproxy.go
package main

import (
	"context"
	"fmt"
	"os"
	"path"
	"strings"
	"sync"
	"text/template"
	"time"

	"github.com/uber-go/zap"
)

const HAPROXY_TPL = `
global
  maxconn {{.MaxConn}}

defaults
  mode http
  maxconn 1024
  option  httplog
  option  dontlognull
  retries 3
  timeout connect 5s
  timeout client  30s
  timeout server  30s

{{ if .EnableStats }}
listen stats
  bind            :{{.StatsPort}}
  mode            http
  maxconn 10
  timeout client  100s
  timeout server  100s
  timeout connect 100s
  timeout queue   100s
  stats enable
  stats hide-version
  stats refresh 30s
  stats show-node
  stats uri /haproxy?stats
{{ end }}

listen circuits
  bind *:{{.Port}}
  mode tcp
  option tcplog
  balance roundrobin

  {{ range $port, $be := .Backends }}
  server tor-{{ $port }} 127.0.0.1:{{ $port }} check maxconn 32{{ end }}
`

// HAProxy helps manage an instance of HAProxy.
type HAProxy struct {
	log zap.Logger
	cmd *Cmd

	dir      string
	conf     string
	template *template.Template
	mu       sync.Mutex
	delay    *time.Timer
	reloadQ  chan bool

	EnableStats bool
	MaxConn     int
	PidFile     string
	Port        int
	StatsPort   int
	Backends    map[int]struct{}
}

func NewHAProxy(ctx context.Context, port int) (h *HAProxy, err error) {
	h = &HAProxy{
		log:     log.With(zap.String("service", "haproxy"), zap.Int("port", port)),
		dir:     "/tmp/torotator/haproxy",
		delay:   time.NewTimer(2 * time.Second),
		reloadQ: make(chan bool, 1),

		EnableStats: *statsPort > 0,
		MaxConn:     256,
		Port:        port,
		StatsPort:   *statsPort,
		Backends:    make(map[int]struct{}),
	}

	t := template.New("haproxy")
	if h.template, err = t.Parse(HAPROXY_TPL); err != nil {
		h.log.Error("unable to parse template", zap.Error(err))
		return
	}

	h.conf = path.Join(h.dir, "haproxy.cfg")
	h.PidFile = path.Join(h.dir, "haproxy.pid")

	if err = h.WriteConfig(ctx, false); err != nil {
		h.log.Error("failed to write config", zap.Error(err))
		return nil, err
	}

	h.cmd, err = NewCommand(ctx, h.log, "haproxy", "-f", h.conf)
	if err != nil {
		h.log.Error("failed to setup command", zap.Error(err))
		return nil, err
	}

	h.cmd.transformLog = h.HAProxyLogger

	return h, nil
}

// MakeDirs attempts to create the directory where HAProxy-related files will reside.
func (h *HAProxy) MakeDirs() (err error) {
	if err = os.MkdirAll(h.dir, 0755); err != nil {
		return
	}

	return nil
}

// HAProxyLogger processes each message received from HAProxy's stdout and stderr. It attempt to categorize each
// message with the correct logging level based on the content of the log line.
func (h *HAProxy) HAProxyLogger(line string) (level, msg string, fields []zap.Field) {
	line = line[1:]

	lvlPos := strings.Index(line, "]")
	level = strings.ToLower(line[:lvlPos])
	switch level {
	case "alert":
		level = "error"
	case "warning":
		level = "warn"
	default:
		h.log.Debug("noticed unmapped log level", zap.String("name", level))
	}

	line = line[lvlPos:]
	msgPos := strings.Index(line, ":")
	msg = line[msgPos+2:]

	return
}

// WriteConfig persists the current HAProxy configuration to disk. This may also signal the current instance of HAProxy
// to reload the configuration after it's written to disk.
func (h *HAProxy) WriteConfig(ctx context.Context, reload bool) (err error) {
	var f *os.File

	if err = h.MakeDirs(); err != nil {
		return
	}

	if f, err = os.OpenFile(h.conf, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644); err != nil {
		return
	}
	defer f.Close()

	h.mu.Lock()
	err = h.template.Execute(f, h)
	h.mu.Unlock()

	if err != nil {
		h.log.Error("unable to render template", zap.Error(err))
		return
	}

	if reload {
		if err = h.Reload(ctx); err != nil {
			h.log.Error("failed to gracefully reload", zap.Error(err))
			return
		}
	}

	return nil
}

// Reload instructs the current instance of HAProxy to finish serving requests, after which a new instance will replace
// it using the newest configuration. This function attempts to throttle requests to reload HAProxy, as many
// Tor+Privoxy pairs may expire at roughly the same time.
func (h *HAProxy) Reload(ctx context.Context) (err error) {
	if !h.delay.Stop() {
		select {
		case <-h.delay.C:
			// drain channel, jic
		default:
		}
	}

	// delay reload for 2 more seconds
	h.delay.Reset(2 * time.Second)
	select {
	case h.reloadQ <- true:
		h.log.Debug("reload queued")
	default:
		h.log.Debug("reload already queued")
		return
	}

	// if we get this far, empty queue when we leave this function. This indicates that a reload may be re-queued.
	defer func() {
		<-h.reloadQ
	}()

	// wait for the timer to expire
	select {
	case <-h.delay.C:
		h.delay.Stop()

	case <-time.After(10 * time.Second):
		// safety net in case we get into a weird state
		return
	}

	prev := h.cmd

	args := []string{"-f", h.conf}
	if prev != nil && prev.cmd != nil {
		args = append(args, "-sf", fmt.Sprintf("%d", prev.Pid()))
	}

	// start a new instance of HAProxy that should allow the current instance to finish up nicely before the new
	// instance takes over
	h.cmd, err = NewCommand(ctx, h.log, "haproxy", args...)
	if err != nil {
		h.log.Error("failed to start new instance", zap.Error(err))
		return
	}

	// try to not leave zombies
	if err = prev.Close(); err != nil {
		h.log.Warn("failed to clean up previous instance", zap.Error(err))
	}

	return nil
}

// AddBackend tells HAProxy that a new Tor+Privoxy backend is available for use.
func (h *HAProxy) AddBackend(ctx context.Context, port int) {
	h.mu.Lock()
	h.Backends[port] = struct{}{}
	h.mu.Unlock()

	h.WriteConfig(ctx, true)
}

// RemoveBackend tells HAProxy that a Tor+Privoxy backend has expired and should be removed from the pool.
func (h *HAProxy) RemoveBackend(ctx context.Context, port int) {
	h.mu.Lock()
	delete(h.Backends, port)
	h.mu.Unlock()

	h.WriteConfig(ctx, true)
}

func (h *HAProxy) Done() <-chan struct{} {
	return h.cmd.Done()
}

func (h *HAProxy) Wait() {
	h.cmd.Wait()
}

func (h *HAProxy) Close() (err error) {
	if h == nil || h.cmd == nil {
		return nil
	}

	defer func() {
		if err = os.RemoveAll(h.dir); err != nil {
			h.log.Error("failed to data directory", zap.String("path", h.dir), zap.Error(err))
		}
	}()

	h.cmd.log.Info("cleaning up")
	if err = h.cmd.Close(); err != nil {
		if err.Error() != "signal: killed" {
			h.cmd.log.Error("failed to kill server", zap.Error(err))
		}
		return err
	}

	return nil
}


cmd/privoxy.go
package main

import (
	"context"
	"fmt"
	"os"
	"path"
	"strings"
	"time"

	"github.com/uber-go/zap"
)

const PRIVOXY_TPL = `
user-manual /usr/share/doc/privoxy/user-manual/
confdir /etc/privoxy
logdir %s
actionsfile match-all.action # Actions that are applied to all sites and maybe overruled later on.
actionsfile default.action   # Main actions file
actionsfile user.action      # User customizations
filterfile default.filter
filterfile user.filter      # User customizations
logfile logfile
listen-address  127.0.0.1:%d
forward-socks5t / 127.0.0.1:%d .
toggle  1
enable-remote-toggle  0
enable-remote-http-toggle  0
enable-edit-actions 0
enforce-blocks 0
buffer-limit 4096
enable-proxy-authentication-forwarding 0
forwarded-connect-retries  0
accept-intercepted-requests 1
allow-cgi-request-crunching 0
split-large-forms 0
keep-alive-timeout 5
tolerate-pipelining 1
socket-timeout 300
`

type Privoxy struct {
	log  zap.Logger
	tor  *Tor
	cmd  *Cmd
	port int
	dir  string
	pid  string
	conf string
}

func NewPrivoxy(ctx context.Context, tor *Tor) (p *Privoxy, err error) {
	p = &Privoxy{tor: tor}

	// loop until we find a port we like
	for {
		select {
		case <-ctx.Done():
			return nil, fmt.Errorf("application terminating")
		default:
		}

		p.port = portPlz()
		p.log = log.With(zap.String("service", "privoxy"),
			zap.Int("port", p.port),
			zap.Int("tor", tor.port))

		p.dir = fmt.Sprintf("/tmp/torotator/privoxy-%d", p.port)
		p.pid = path.Join(p.dir, "privoxy.pid")
		p.conf = path.Join(p.dir, "privoxy.conf")

		if err = p.WriteConfig(); err != nil {
			p.log.Error("failed to write config", zap.Error(err))
			continue
		}

		p.cmd, err = NewCommand(ctx, p.log, "privoxy",
			"--no-daemon",
			"--pidfile", p.pid,
			p.conf)
		if err != nil {
			p.log.Error("failed to setup command", zap.Error(err))
			time.Sleep(500 * time.Millisecond)
			continue
		}

		p.cmd.transformLog = p.PrivoxyLogger

		break
	}

	return p, nil
}

func (p *Privoxy) WriteConfig() (err error) {
	if err = os.MkdirAll(p.dir, 0755); err != nil {
		return
	}

	var f *os.File
	if f, err = os.OpenFile(p.conf, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644); err != nil {
		return
	}
	defer f.Close()

	f.WriteString(fmt.Sprintf(PRIVOXY_TPL, p.dir, p.port, p.tor.port))

	return nil
}

func (p *Privoxy) PrivoxyLogger(line string) (level, msg string, fields []zap.Field) {
	line = line[37:]

	lvlPos := strings.Index(line, ":")
	level = strings.ToLower(line[:lvlPos])
	if strings.Contains(level, " ") {
		level = strings.Split(level, " ")[0]
	}

	msg = line[lvlPos+2:]

	return
}

func (p *Privoxy) Done() <-chan struct{} {
	return p.cmd.Done()
}

func (p *Privoxy) Wait() {
	p.cmd.Wait()
}

func (p *Privoxy) Close() (err error) {
	if p == nil {
		return nil
	}

	defer func() {
		if err = os.RemoveAll(p.dir); err != nil {
			p.log.Error("failed to data directory", zap.String("path", p.dir), zap.Error(err))
		}
	}()

	p.cmd.log.Info("cleaning up")
	if err = p.cmd.Close(); err != nil {
		if err.Error() != "signal: killed" {
			p.cmd.log.Error("failed to kill server", zap.Error(err))
		}
		return err
	}

	return nil
}


cmd/torotator.go
package main

import (
	"context"
	"flag"
	"os"
	"os/exec"
	"os/signal"
	"sync"
	"syscall"
	"time"

	"github.com/uber-go/zap"
)

var (
	VERSION = "dev"

	proxyPort      = flag.Int("p", 8080, "HTTP proxy port")
	torCount       = flag.Int("c", 3, "number of Tor nodes to use")
	portRangeStart = flag.Int("s", 30000, "starting port for proxy usage")
	maxProxyTime   = flag.Int("m", 900, "maximum time (in seconds) a proxy should remain online before being recycled")
	circuitTime    = flag.Int("t", 120, "maximum time (in seconds) a Tor node should be online before recircuiting")
	statsPort      = flag.Int("stats", 0, "serve HAProxy stats on this port")
	debug          = flag.Bool("debug", false, "enable debug mode")
	version        = flag.Bool("v", false, "show version and exit")

	log zap.Logger
)

func init() {
	flag.Parse()

	log = zap.New(zap.NewJSONEncoder(zap.RFC3339Formatter("time")))
	if *debug {
		log.SetLevel(zap.DebugLevel)
	}

	log.Info("rotating tor proxy", zap.String("version", VERSION))
	if *version {
		os.Exit(0)
	}
}

func main() {
	FindDependencies()

	ctx := SignalContext()
	wg := new(sync.WaitGroup)

	ha, err := NewHAProxy(ctx, *proxyPort)
	if err != nil {
		log.Fatal("failed to start HAproxy", zap.Error(err))
	}

	defer ha.Close()
	go ha.Wait()
	go ReloadOnHUP(ctx, ha)

	Rotate(ctx, wg, ha)

	// clean up
	wg.Wait()
	log.Info("done")
}

func FindDependencies() {
	var (
		found string
		err   error
	)

	deps := []string{"haproxy", "tor"}
	for _, dep := range deps {
		if found, err = exec.LookPath(dep); err != nil {
			log.Fatal("missing required program", zap.String("name", dep))
		} else {
			log.Debug("found required program", zap.String("name", dep), zap.String("path", found))
		}
	}
}

// Rotate manages pairs of Tor+Privoxy services. Only a specific number of pairs are permitted at one time. When a pair
// expires, a new pair will automatically take its place.
func Rotate(ctx context.Context, wg *sync.WaitGroup, ha *HAProxy) {
	// Used to limit the number of running proxies. This is separate from wg because wg is unbounded.
	c := make(chan bool, *torCount)

	for {
		select {
		case <-ctx.Done():
			// application terminating
			close(c)
			return

		default:
			c <- true

			// time to create a new pair
			wg.Add(1)
			go func() {
				RunProxy(ctx, ha)

				wg.Done()
				<-c
			}()
		}
	}
}

// RunProxy creates a Tor node and notifies HAProxy so it can reconfigure itself to use the new pair. If the Tor node
// fails, the circuit is invalidated and removed from HAProxy.
func RunProxy(ctx context.Context, ha *HAProxy) {
	tor, err := NewTor(ctx)
	if err != nil {
		tor.Close()
		return
	}

	_log := log.With(zap.Int("tor", tor.port))
	_log.Info("proxy started")

	// notify HAProxy of the new backend
	ha.AddBackend(ctx, tor.port)

	// let the processes run until they terminate
	go tor.Wait()

	// TODO periodically check that this proxy is still functional
	// wait for any of the following events to occur
	select {
	case <-ctx.Done():
		// application terminating
	case <-tor.Done():
		// tor ended
	case <-time.After(time.Duration(*maxProxyTime) * time.Second):
		// proxy lifetime expired
	}

	// tell HAProxy to remove this backend
	ha.RemoveBackend(ctx, tor.port)

	// clean up after ourselves
	_log.Info("stopping proxy")
	tor.Close()

	_log.Info("proxy terminated")
}

// SignalContext creates a new context that will be canceled when the program receives certain termination signals.
func SignalContext() context.Context {
	ctx, cancel := context.WithCancel(context.Background())

	// handle termination signals
	terminate := make(chan os.Signal, 1)
	signal.Notify(terminate, os.Kill, os.Interrupt)

	go func() {
		<-terminate
		cancel()
	}()

	return ctx
}

// ReloadOnHUP waits to receive a SIGHUP signal, at which point HAProxy will reload its configuration.
func ReloadOnHUP(ctx context.Context, ha *HAProxy) {
	hup := make(chan os.Signal, 1)
	signal.Notify(hup, syscall.SIGHUP)

	go func() {
		for _ = range hup {
			log.Info("got sighup; reloading config")
			ha.Reload(ctx)
		}
	}()
}
